### ISSUE TITLE
VPN tunnel connected but data stops flowing

### ISSUE THREAD

**Issue by jdanders** [schwabe/ics-openvpn#777]
# General information
1. Android 7.1.2
2. Android Vendor: LineageOS
3. Device: Moto X 2014
4. Version of the app: 0.6.73 from Play Store
 
# Description of the issue
I have been running OpenVPN between my phone and home server for many years. I occasionally see the VPN connection get into a state where the client thinks it is connected, but no data flows. To automatically fix this, I have a Tasker job that periodically tries to fetch HTTP through the VPN and if it fails, forcibly restarts the OpenVPN client. Since I updated to Android 7 I am seeing this happen more often.

I decided to take a closer look and see if I can diagnose something. This is a hard one to track, because it only happens after a decent amount of time and the client reports **nothing** in the logs, even on max verbosity. The last message shown while bad is from the last connection: `Debug state info: CONNECTED to WIFI "myap", pause: userPause, shouldbeconnected: true, network: SHOULDBECONNECTED`. So that told me I needed to capture server logs for long enough to trace back through and see what happened.

I finally did last night, and the symptom I saw was that every time the phone sent a packet through the tun0 interface while in this state, the server dropped the packet and logged a message like: 

`1194 AEAD Decrypt error: cipher final failed`

Google pointed me to an [already existing issue](https://community.openvpn.net/openvpn/ticket/847) in the openvpn tracker. The description of that issue includes this statement:

> A renegotiation is forced (due to --reneg-sec 10). At this point do not enter any credentials until the server have logged:
Sat Feb 25 00:34:55 2017 us=576093 Test-Client/192.168.122.9:1194 TLS Error: TLS key negotiation failed to occur within 60 seconds (check your network connectivity)

This is where I think the Android client might be able to work around this issue and not get the server into this state. **Is it possible that the client takes longer than 60 seconds to use the credentials to renegotiate?** Most likely I would think it's related to Android's aggressive doze mode, but I don't really know.

I am using a "Android Certificate" hardware keystore certificate for my client credentials. These settings in my config might be relevant:
```
ping 35
ping-restart 0
ncp-disable
connect-retry 2 90
resolv-retry 60
```
One final note: I don't know that the linked issue is the one I am running into, because in that issue they describe needing to restart the server to get things up and going again, where I know that restarting the client is sufficient for me. However, that might be because I have the server set to drop active sessions if a client creates a new connection (but I think that's default behavior and therefore shouldn't be different than the bug report). I am trying to capture this state with full verbosity logs on the server, but it hasn't happened yet. I will update if I learn anything new.

**Comment by schwabe** [comment]
I have no idea what is happening with your client. You can adjust the tls-timeout (not 100% about the option) setting to change the renogiation timeout.

**Comment by jdanders** [comment]
I have an issue on the server software: https://community.openvpn.net/openvpn/ticket/950

Right now it looks like maybe the Android client is opening two simultaneous connections to the server, and since they share the same CN, bad things happen. Is there some reason that the client would open two connections, and then even worse keep them both open?

**Comment by schwabe** [comment]
The Android client is the same as normal openvpn. It does not open two connections. But it does roam with peer-id and does instant reconnect without peer-id. This timing is different than most desktop clients and that might upset something in the OpenVPN code (although it shouldn't). Of the reconnection has the same connection. The old connection might not be timed out but that should not matter.

**Comment by jdanders** [comment]
Is there a reason it changes peer-id? I see this in the log I submitted in the openvpn issue:
```
10:15:01 2017 us=127647 phone/192.168.1.1:40119 SENT CONTROL [phone]: 'PUSH_REPLY,redirect-gateway autolocal bypass-dhcp,route 192.168.1.0 255.255.255.0,dhcp-option DNS 192.168.1.1,route 192.168.8.1,topology net30,ifconfig 192.168.8.6 192.168.8.5,peer-id 0' (status=1)
11:57:04 2017 us=467161 phone/192.168.1.1:36571 SENT CONTROL [phone]: 'PUSH_REPLY,redirect-gateway autolocal bypass-dhcp,route 192.168.1.0 255.255.255.0,dhcp-option DNS 192.168.1.1,route 192.168.8.1,topology net30,ifconfig 192.168.8.6 192.168.8.5,peer-id 1' (status=1)
16:10:04 2017 us=580133 phone/192.168.1.1:59790 SENT CONTROL [phone]: 'PUSH_REPLY,redirect-gateway autolocal bypass-dhcp,route 192.168.1.0 255.255.255.0,dhcp-option DNS 192.168.1.1,route 192.168.8.1,topology net30,ifconfig 192.168.8.6 192.168.8.5,peer-id 0' (status=1)
16:37:27 2017 us=520940 phone/192.168.1.1:52655 SENT CONTROL [phone]: 'PUSH_REPLY,redirect-gateway autolocal bypass-dhcp,route 192.168.1.0 255.255.255.0,dhcp-option DNS 192.168.1.1,route 192.168.8.1,topology net30,ifconfig 192.168.8.6 192.168.8.5,peer-id 1' (status=1)
```
The changes at 10:15, 11:57, and 16:10 worked fine. The problem seems to be at 16:37, the connection at 16:10 didn't die, so both peer-id 0 (tcp port 59790) and peer-id 1 (tcp port 52655) tried to remain active.

What would cause the Android client to start using a new connection and new peer-id?

### Edit:
I saw this happen in my current session and it is preceeded with the phone client saying:
```
write UDP[]: Network is unreachable (code=101)
```
Looking at more vebosity in the log, it seems that the phone was momentarily disconnected from the wifi. That message was seen at various times during the night, but at 7:07 this morning, it repeated 19 times and was followed by a `SIGTERM[soft,management-exit] receieved, process exiting`. It then immediately started over and made a new connection. 

Looking at higher verbosity in the log I see that at 7:07 the network switched from disconnected to LTE and back to Wifi very quickly. Between LTE and Wifi, there is a `PID_ERR replay-window backtrack occurred` message. After the switch back to wifi, I see:
```
Debug state info: CONNECTED to WIFI "myap", pause: userPause, shouldbeconnected: true, network: SHOULDBECONNECTED
Debug state info: CONNECTED to WIFI "myap", pause: userPause, shouldbeconnected: true, network: SHOULDBECONNECTED
Building configuration...
```
I'm not sure why I don't see `SIGTERM` in the verbose logs.

This reconnect did not result in any problems after the restart.

Assuming this is expected behavior, it seems as though in my problem log (when there was a problem after a switch) the old session did not die, but a new connection was created.

**Comment by jdanders** [comment]
I ran a log that captures when things went bad. While bad I confirmed that there are two connections open on the client:
```
netstat -anp|grep vpn
udp 0 0 0.0.0.0:37755 0.0.0.0:* 5051/c_pie_openvpn.armeabi-v7a
udp 0 0 0.0.0.0:40916 0.0.0.0:* 2455/c_pie_openvpn.armeabi-v7a
unix 2 [ ACC ] STREAM LISTENING 5206625 6168/de.blinkt.open/data/user/0/de.blinkt.openvpn/cache/mgmtsocket
unix 2 [ ACC ] STREAM LISTENING 5221764 6168/de.blinkt.open/data/user/0/de.blinkt.openvpn/cache/mgmtsocket
unix 3 [ ] STREAM CONNECTED 5206690 2455/c_pie_openvpn.
unix 3 [ ] STREAM CONNECTED 5207198 6168/de.blinkt.open/data/user/0/de.blinkt.openvpn/cache/mgmtsocket
unix 3 [ ] STREAM CONNECTED 5220230 6168/de.blinkt.open/data/user/0/de.blinkt.openvpn/cache/mgmtsocket
unix 3 [ ] STREAM CONNECTED 5223685 5051/c_pie_openvpn.
```
Things start being bad at 19:08. In the client log I see this:
```
2017-10-24 19:08:59 Ignoring OpenVPN Status in CONNECTED state (WAIT->LEVEL_CONNECTING_NO_SERVER_REPLY_YET): ,,,,,
2017-10-24 19:08:59 Ignoring OpenVPN Status in CONNECTED state (AUTH->LEVEL_CONNECTING_SERVER_REPLIED): ,,,,,
```
I'm not sure what that means. I will attach the client log. Looks like timestamps mostly match the server logs attached in the [openvpn issue](https://community.openvpn.net/openvpn/ticket/950).
[vpn_client.log](https://github.com/schwabe/ics-openvpn/files/1413162/vpn_client.log)

**Comment by jdanders** [comment]
I turned --ping-restart back on in my config and now things are even stranger. I still get to the state when the client opens two connections, and the VPN stops working because of it. But now, I get a ping-restart according to the timeout I set, which causes the client to renegotiate, as expected. When the negotiation is done, data can flow through the VPN, **but there are still two connections open!** And even more strange is that in spite of a working VPN, the client continues to ping-restart every timeout time later until I actually do a reconnect in the GUI.
